viewing_party/party.py

```diff
@@ -1,23 +1,139 @@
+    if num == 0:
+        return 0
+    else:
+        for i in range(num):
+            score += user_data['watched'][i]['rating']
+        return score/num
+
+# test 3&4
+
+
+def get_most_watched_genre(user_data):
+    num = len(user_data['watched'])
+    genre_dict = {}
+    if num == 0:
+        return None
+    else:
+        for i in range(num):
+            if user_data['watched'][i]['genre'] not in genre_dict:
+                genre_dict[user_data['watched'][i]['genre']] = 0
+            elif user_data['watched'][i]['genre'] in genre_dict:
+                genre_dict[user_data['watched'][i]['genre']] += 1
+        max_value = max(genre_dict, key=genre_dict.get)
+    return max_value
 
 # -----------------------------------------
 # ------------- WAVE 3 --------------------
 # -----------------------------------------
 
-        
+
+def get_unique_watched(user_data):
+    unique_list = []
+    movie_list = []
+    num = len(user_data['friends'])
+    for i in range(num):
+        movie_list += user_data['friends'][i]['watched']
+    for movie_dict in user_data['watched']:
+        if movie_dict not in movie_list:
+            unique_list.append(movie_dict)  # append dict, not +=
+    return unique_list
+
+
+def get_friends_unique_watched(user_data):
+    unique_list = []
+    movie_list = []
+    num = len(user_data['friends'])
+    for i in range(num):
+        movie_list += user_data['friends'][i]['watched']
+    for movie_dict in movie_list:
+        if movie_dict not in unique_list:
+            if movie_dict not in user_data['watched']:
+                unique_list.append(movie_dict)  # append dict, not +=
+    return unique_list
+# question about assertion in test wave 03
+
 # -----------------------------------------
 # ------------- WAVE 4 --------------------
 # -----------------------------------------
 
+
+def get_available_recs(user_data):
+    movie_recs = []
+    fr_unique_list = get_friends_unique_watched(user_data)
+    for movie_dict in fr_unique_list:
+        if movie_dict['host'] in user_data['subscriptions']:
+            movie_recs.append(movie_dict)
+    return movie_recs
 # -----------------------------------------
 # ------------- WAVE 5 --------------------
 # -----------------------------------------
 
+
+def get_new_rec_by_genre(user_data):
+    movie_recs = []
+    fr_unique_list = get_friends_unique_watched(user_data)
+    max_genre = get_most_watched_genre(user_data)
+    for movie_dict in fr_unique_list:
+        if movie_dict['genre'] == max_genre:
+            movie_recs.append(movie_dict)
+    return movie_recs
+
+
+def get_rec_from_favorites(user_data):
+    fav_list = []
+    movie_list = []
+    num = len(user_data['friends'])
+    for i in range(num):
+        movie_list += user_data['friends'][i]['watched']
+    for movie_dict in user_data['favorites']:
+        if movie_dict not in movie_list:
+            fav_list.append(movie_dict)  # append dict, not +=
+    return fav_list

```

Comment:
Nice approach! Another possibility could involve reusing our function `get_unique_watched`:

```python
my_unique_watched = get_unique_watched(user_data)
recommended = []

for movie in user_data['favorites']:
    if movie in my_unique_watched:
        recommended.append(movie)

return recommended
```

Reply:
Kelsey-Thank you for reviewing my project and providing many helpful feedback! I am glad you saw a pattern of "bad habits" so that I can change what I am used to. I typed out your suggestions in VS code and tried different things out.  I'll incorporate your suggestions and try to be more pythonic and concise next time! TY again!